Moved torch headers copy to build_deps

sdmonov opened the pull request:
PR #5706 initially moved headers under build_ext to fix bdist_wheel and
build develop. This broke install and #5755 moved them back to install
which broke bdist_wheel and build develop. Looks like build_ext is called
from install after it already tried to copy the headers to the python install
dir and the headers were not installed correctly. Using build_deps works
correct with all setup.py install, bdist_wheel and build develop.

@goldsborough

sdmonov committed:
Moved torch headers copy to build_deps

PR #5706 initially moved headers under build_ext to fix bdist_wheel and
build develop. This broke install and #5755 moved them back to install
which broke bdist_wheel and build develop. Looks like build_ext is called
from install after it already tried to copy the headers to the python install
dir and the headers were not installed correctly. Using build_deps works
correct with all setup.py install, bdist_wheel and build develop.

apaszke reviewed (commented):
I don’t think this is correct. It might work on your system because you already had the auto generated files from previous builds, but these commands would now run before the codegen (which is in build_ext)

goldsborough commented:
Yes please be sure to do a `find . -name 'torch.h'` and remove any torch.h files from previous builds(`torch/lib/include`) before you test this. I've been tripped up by this a few times too.

goldsborough commented:
That said, the CI looks green, so it seems to work.

apaszke commented:
I’m sure that this runs before because build_deps is a dependency of build_ext. Not sure how can this succeed, but maybe our CI isnt using an entirely fresh every time either

goldsborough commented:
I actually just tried it locally and it works for both `build develop` and `install`. Our CI starts from a fresh ubuntu docker image so it's pretty unbiased. Looking at `setup.py`, this looks legit: `build_deps` is called directly from `install` and `build`

apaszke commented:
Ohhhh ok I think I know what's happening. It's probably the case that the build doesn't depend on any of the auto-generated code, so we're just shipping a slightly incomplete, but sufficient subset of our `csrc`.

goldsborough commented:
Yes that's why I had problems getting `generated/VariableType.h` the other day. So this solution works right now, but it's not super perfect

apaszke commented:
Yup. I think it's good to merge, but let's add a comment saying that if you want to use the auto-generated files, then you'll need to figure out a better way

sdmonov commented:
I have been testing this with clean environments for the different cases. Like @goldsborough mention `build_deps` is called directly from `install` and `build` and that's why I put the copy there.

ezyang reviewed (changes requested):
More commentary on the `copy_tree` about the limitations of the approach, as per PR discussion

sdmonov committed:
Comment about the auto-generated files

Added comment that the current solution will not include auto-generated
files which may be a problem if somebody needs to use them

sdmonov commented:
@pytorchbot test this please

prigoyal commented:
Hi all, this PR addresses the issue #5964 , can we do the review again ? :) thanks a lot.
